File: germanium/points/box.js
//noinspection JSAnnotator,JSReferencingArgumentsOutsideOfFunction,JSUnresolvedVariable,ThisExpressionReferencesGlobalObjectJS
return (function() {
    var style = document.body.currentStyle ||
                window.getComputedStyle(document.body);

    var bodyMarginTop = parseInt(style.marginTop),
        bodyMarginLeft = parseInt(style.marginLeft);

    function left(element) {
        var result = - bodyMarginLeft;
        while (element && element != document.body) {
            result += element.offsetLeft;
            element = element.offsetParent;
        }

        return result;
    }

    function top(element) {
        var result = - bodyMarginTop;
        while (element && element != document.body) {
            result += element.offsetTop;
            element = element.offsetParent;
        }

        return result;
    }

    function right(element) {
        var result = element.offsetWidth - bodyMarginLeft;
        while (element && element != document.body) {
            result += element.offsetLeft;
            element = element.offsetParent;
        }

        return result;
    }

    function bottom(element) {
        var result = element.offsetHeight - bodyMarginTop;
        while (element && element != document.body) {
            result += element.offsetTop;
            element = element.offsetParent;
        }

        return result;
    }

    var element = arguments[0];

    var _top = top(element),
        _right = right(element),
        _bottom = bottom(element),
        _left = left(element),
        _width = _right - _left,
        _height = _bottom - _top;

    // the right and the bottom values need to be adjusted to be inside
    // the box.
    
    _right -= 1;
    _bottom -= 1;

    return [_top, _right, _bottom, _left, _width, _height];
}.apply(this, arguments));

File: germanium/points/Box.py
from germanium.impl._load_script import load_script
from .Point import Point


class Box(object):
    def __init__(self, selector):
        self._selector = selector
        self._box = None

    def top_left(self):
        if not self._box:
            self.get_element_box()

        return Point(self._box['left'],
                     self._box['top'])

    def top_right(self):
        if not self._box:
            self.get_element_box()

        return Point(self._box['right'],
                     self._box['top'])

    def bottom_left(self):
        if not self._box:
            self.get_element_box()

        return Point(self._box['left'],
                     self._box['bottom'])

    def bottom_right(self):
        if not self._box:
            self.get_element_box()

        return Point(self._box['right'],
                     self._box['bottom'])

    def get_element_box(self):
        from germanium.static import S, js
        code = load_script(__name__, 'box.min.js')

        top, right, bottom, left, width, height = js(code, S(self._selector).element())

        self._box = {
            "top": top,
            "right": right,
            "bottom": bottom,
            "left": left,
            "width": width,
            "height": height
        }

        print("found box %s " % self._box)

        return self._box
